aip/general/0193.md

```diff
@@ -210,41 +280,237 @@ map for the error payload to be backwards compatible, even if the value
+        time.`
+  - If the error resolution exceeds the number of characters supported
+    in the problem description (`message` string of
+    `google.rpc.Status`), or requires multiple troubleshooting steps,
+    include the most common resolution in the message and use the
+    `HelpLink` payload to link to relevant documentation.
+
+##### HelpLink payload
+
+When `LocalizedMessage.message` doesn’t provide the user sufficient
+context or actionable next steps, or if there are multiple points of
+failure that need to be considered in troubleshooting, a link to
+supplemental troubleshooting documentation **must** be provided in the
+`HelpLink`payload.
+
+Provide this information in addition to a clear problem definition and
+actionable resolution, not as an alternative to them; for example.
+
+**Note:** For more information, see the troubleshooting documentation:
+[https://cloud.google.com/compute/docs/resource-error][resource-error].
+
+[HelpLink payload][HelpLink]
+
+Details of this object are summarized in the following ields, field
+descriptions, and examples:
+
+- `description string`
+  - Describes what the link offers.
+  - *Example*: "Troubleshooting documentation for STOCKOUT errors"
+- `url string`
+  - The URL of the link. For error documentation this must follow the
+    standardized format listed in the following example.
+  - *Example*:
+    https://cloud.google.com/PRODUCT/docs/ERROR-REASON
+
+
+### Error messages
+
+The following sections describe best practices for error messages not
+described previously.
+
+
+#### Changing error messages
+
+
+##### Method
+
+The method of changing error messages depends on whether an ErrorInfo
+payload is present in the message.
+
+If the error message contains ErrorInfo payload (a machine-readable
+identifier):
+
+- `Status.message` *string* and `LocalizedMessage.message` *string* can change
+- New metadata fields can be added
+- However, existing metadata fields **must not** be removed and existing
+  metadata field keys cannot be modified.
+
+If the error message does *not* contain ErrorInfo payload (usually for
+pre-existing APIs):
+
+- The `LocalizedMessage.message` *string* can change
+- However, the `Status.message` *string* **must** not be changed, as
+  this change is backward-incompatible.
+
+##### Message alignment
+
+By default, `LocalizedMessage` is populated with the same message as
+`Status.message`. Should you choose, you can also present a different
+`LocalizedMessage.message` from `Status.message`. The following are some
+reasons why.
+
+Keeping the aforementioned guidance in mind, the safest and recommended
+method to update an error message for a service is to add
+[`google.rpc.LocalizedMessage`][LocalizedMessage] to
+[`Status.details`][Status details].  `LocalizedMessage` is meant for
+displaying messages to end users.
+
+The error message presented in `LocalizedMessage.message` **should** be
+the same as `Status.message` or it **should** be an alternate message.
```

Comment:
This is a bit confusing it should be one or it should be another. This doesn 't seem logical.

Reply:
+1. This was confusing to me too. The statement seems redundant. I'd vote for removal.

Reply:
Done.